fido2/tpm.py

```diff
@@ -0,0 +1,308 @@
+# Copyright (c) 2019 Yubico AB
+# All rights reserved.
+#
+#   Redistribution and use in source and binary forms, with or
+#   without modification, are permitted provided that the following
+#   conditions are met:
+#
+#    1. Redistributions of source code must retain the above copyright
+#       notice, this list of conditions and the following disclaimer.
+#    2. Redistributions in binary form must reproduce the above
+#       copyright notice, this list of conditions and the following
+#       disclaimer in the documentation and/or other materials provided
+#       with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import struct
+from io import BytesIO
+from binascii import b2a_hex
+
+from enum import IntEnum, IntFlag
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.asymmetric import rsa
+
+from .utils import bytes2int
+
+
+TPM_ALG_NULL = 0x0010
+
+
+class TpmAlgAsym(IntEnum):
+    RSA = 0x0001
+    ECC = 0x0023
+
+
+class TpmAlgHash(IntEnum):
+    SHA1 = 0x0004
+    SHA256 = 0x000B
+    SHA384 = 0x000C
+    SHA512 = 0x000D
+
+
+class TpmAttestationFormat(object):
+    """the signature data is defined by [TPMv2-Part2] Section 10.12.8 (TPMS_ATTEST)
+    as:
+      TPM_GENERATED_VALUE (0xff544347 aka "\xffTCG")
+      TPMI_ST_ATTEST - always TPM_ST_ATTEST_CERTIFY (0x8017)
+        because signing procedure defines it should call TPM_Certify
+        [TPMv2-Part3] Section 18.2
+      TPM2B_NAME
+        size (uint16)
+        name (size long)
+      TPM2B_DATA
+        size (uint16)
+        name (size long)
+      TPMS_CLOCK_INFO
+        clock (uint64)
+        resetCount (uint32)
+        restartCount (uint32)
+        safe (byte) 1 yes, 0 no
+      firmwareVersion uint64
+      attested TPMS_CERTIFY_INFO (because TPM_ST_ATTEST_CERTIFY)
+        name TPM2B_NAME
+        qualified_name TPM2B_NAME
+    See:
+      https://www.trustedcomputinggroup.org/wp-content/uploads/TPM-Rev-2.0-Part-2-Structures-01.38.pdf
+      https://www.trustedcomputinggroup.org/wp-content/uploads/TPM-Rev-2.0-Part-3-Commands-01.38.pdf
+    """
+
+    TPM_GENERATED_VALUE = b"\xffTCG"
+    TPM_ST_ATTEST_CERTIFY = b"\x80\x17"
```

Comment:
I think this would be more readable if these values were ints specified as hex, eg:
`TPM_GENERATED_VALUE = 0xff544347`.

Reply:
TCG stands for trusted computing group, I fought it made more sense.

Reply:
Ah, I see. Yeah, that makes sense.